lib/DAV/LockPlugin.php

```diff
@@ -0,0 +1,229 @@
+		});
+
+		$propFind->handle(Application::DAV_PROPERTY_LOCK_OWNER_TYPE, function () use ($nodeId) {
+			$lock = $this->lockService->getLockForNodeId($nodeId);
+
+			if ($lock === false) {
+				return null;
+			}
+
+			return $lock->getType();
+		});
+
+		$propFind->handle(Application::DAV_PROPERTY_LOCK_EDITOR, function () use ($nodeId) {
+			$lock = $this->lockService->getLockForNodeId($nodeId);
+			if ($lock === false || $lock->getType() !== ILock::TYPE_APP) {
+				return null;
+			}
+
+			return $lock->getOwner();
+		});
+
+		$propFind->handle(Application::DAV_PROPERTY_LOCK_TOKEN, function () use ($nodeId) {
+			$lock = $this->lockService->getLockForNodeId($nodeId);
+			if ($lock === false) {
+				return null;
+			}
+
+			return $lock->getToken();
+		});
+	}
+
+	private function getLockDisplayName(ILock $lock): ?string {
+		$user = $this->userManager->get($lock->getOwner());
+		if ($user !== null) {
+			return $user->getDisplayName();
+		}
+
+		return null;
+	}
+
+	public function httpLock(RequestInterface $request, ResponseInterface $response) {
+		if ($request->getHeader('X-User-Lock')) {
+			$response->setHeader('Content-Type', 'application/xml; charset=utf-8');
+
+			try {
+				$file = $this->fileService->getFileFromAbsoluteUri($this->server->getRequestUri());
+				$lockInfo = $this->lockService->lock(new LockContext(
+					$file, ILock::TYPE_USER, $this->userSession->getUser()->getUID()
+				));
+				$response->setStatus(200);
+				$response->setBody(
+					$this->server->xml->write(
+						'{DAV:}prop',
+						$this->getLockProperties($lockInfo)
+					)
+				);
+			} catch (OwnerLockedException $e) {
+				$response->setStatus(423);
+				$response->setBody(
+					$this->server->xml->write(
+						'{DAV:}prop',
+						$this->getLockProperties($e->getLock())
+					)
+				);
+			}
+
+			return false;
+		}
+
+		return parent::httpLock($request, $response);
+	}
+
+	public function httpUnlock(RequestInterface $request, ResponseInterface $response) {
+		if ($request->getHeader('X-User-Lock')) {
+			$response->setHeader('Content-Type', 'application/xml; charset=utf-8');
+
+			try {
+				$file = $this->fileService->getFileFromAbsoluteUri($this->server->getRequestUri());
+				$this->lockService->unlock(new LockContext(
+					$file, ILock::TYPE_USER, $this->userSession->getUser()->getUID()
+				));
+				$response->setStatus(200);
+				$response->setBody(
+					$this->server->xml->write(
+						'{DAV:}prop',
+						$this->getLockProperties(null)
+					)
+				);
+			} catch (LockNotFoundException $e) {
+				$response->setStatus(Http::STATUS_PRECONDITION_FAILED);
+			} catch (UnauthorizedUnlockException $e) {
+				$response->setStatus(Http::STATUS_LOCKED);
+			} catch (InvalidPathException $e) {
+			} catch (NotFoundException $e) {
+			}
+
+			return false;
+		}
+
+		return parent::httpUnlock($request, $response);
+	}
+
+	private function getLockProperties(?FileLock $lock): array {
+		return [
+			Application::DAV_PROPERTY_LOCK => (bool)$lock,
```

Comment:
Are you sure this is possible? I don't see any obvious reason in the FileLock class why this would work. :grin:
Or is it equivalent to `=> $lock !== null`? Maybe it's more explicit like that.